### PR TITLE
fixes #6

### LINKED ISSUE #6: We need to complete the ASP.NET Core integration package
We need to complete the ASP.NET Core integration package to get it running in ASP.NET Core and .NET Core

- Needs some more time, because it needs a generic and proper way to move the existing service registrations from the ASP.NET Core's IServiceCollection to LightCore
- The current implementation doesn't catch two pretty special service registrations from the IServiceCollection to LightCore, which are the following:
 
`{System.Collections.Generic.IEnumerable`1[Microsoft.Extensions.Options.IConfigureOptions`1[Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions]] setups}`

`{System.Collections.Generic.IEnumerable`1[Microsoft.Extensions.Options.IPostConfigureOptions`1[Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions]] postConfigures}`

e.g.
```
using Microsoft.Extensions.Options;
using Microsoft.AspNetCore.Server.Kestrel.Core;

IEnumerable<IConfigureOptions<KestrelServerOptions>>
IEnumerable<IPostConfigureOptions<KestrelServerOptions>>
```

### PR DESCRIPTION
**Pull request by JuergenGutsch**
This PR resolves the registration issues in ASP.NET Core

### CODE REVIEW

File: Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+* Threading issue is related to the tests
+  * Test setup is done kinda wrong
+* ASP.NET Core Integration doesn't work because of an missing feature in LightCore
+  * We need to support dynamic mappings
```
Comment (peterbucher) [inline Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md:4]: How should "dynamic mappings" look like? Is there some kind of spec for this?

File: LightCore.Integration.AspNetCore/LightCoreExtension.cs [hunk 2]
```diff
@@ -39,8 +45,19 @@ private static void RegisterServices(IContainerBuilder builder, IServiceCollecti
 
         private static void Register(IContainerBuilder builder, ServiceDescriptor serviceDescriptor, ILifecycle lifecycle)
         {
+
             if (serviceDescriptor.ImplementationType != null)
             {
+                var serviceTypeInfo = serviceDescriptor.ServiceType.GetTypeInfo();
```
Comment (peterbucher) [inline LightCore.Integration.AspNetCore/LightCoreExtension.cs]: What is this suppose to do?

File: LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs [hunk 3]
```diff
@@ -10,13 +10,6 @@ namespace LightCore.Integration.AspNetCore.Lifecycle
     /// </summary>
     public class HttpRequestLifecycle : ILifecycle
     {
-        private readonly IHttpContextAccessor _httpContext;
-
-        public HttpRequestLifecycle(IHttpContextAccessor httpContext)
```
Comment (peterbucher) [inline LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs:15]: The IHttpContextAccessor parameter was here to enable Unittesting of the Lifecycle. Why did you remove it?

File: LightCore/Registration/RegistrationSource/EnumerableRegistrationSource.cs [hunk 4]
```diff
@@ -31,15 +31,8 @@ internal class EnumerableRegistrationSource : IRegistrationSource
...
-                        &&
-                        ((this._registrationContainer.HasRegistration(
-                            contractType.GetGenericArguments().FirstOrDefault()))
-                         ||
-                         // Use ConcreteTypeRegistrationSource.
-                         (this._registrationContainer.IsSupportedByRegistrationSource(
-                             contractType.GetGenericArguments().FirstOrDefault())));
-
+                // Don't check the generic arguments, to ensure to create enumerables anyway.
+                return contractType => contractType.IsGenericEnumerable();
```
Comment (peterbucher) [inline LightCore/Registration/RegistrationSource/EnumerableRegistrationSource.cs:35]: So this is the real fix for ASP.NET Core Support. Because there are Enumerables demanded without any Enumerable Item Type registered within LightCore, right?

File: Samples/LightCore.Integration.AspNetCore.Sample/Startup.cs [hunk 5]
```diff
@@ -26,40 +18,19 @@ public Startup(IConfiguration configuration)
...
-            // {System.Collections.Generic.IEnumerable`1[Microsoft.ApplicationInsights.AspNetCore.ITelemetryProcessorFactory] telemetryProcessorFactories}
-            builder.RegisterFactory<IEnumerable<ITelemetryProcessorFactory>>(x => { return new List<ITelemetryProcessorFactory>(); });
-
-            //{ Void.ctor(
-            //    System.Collections.Generic.IEnumerable`1[Microsoft.Extensions.Options.IConfigureOptions`1[Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.LibuvTransportOptions]], 
-            //    System.Collections.Generic.IEnumerable`1[Microsoft.Extensions.Options.IPostConfigureOptions`1[Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.LibuvTransportOptions]])}
-
-
-            //builder.Register<IService, MyService>();
+            // small hack because this cannot resolved yet. Lightcore wont resolve unknown types
```
Comment (peterbucher) [inline Samples/LightCore.Integration.AspNetCore.Sample/Startup.cs]: How to fix this?

**Review by peterbucher — COMMENTED** [review]
Hey Jürgen. Please explain me what it is abount the removal of "IHttpContextAccessor", all the other things are looking fine. But please answer the comment questions.

Thanks!

File: Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md (hunk 1 above)
Comment (JuergenGutsch) [reply Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md:4]: This is just an old description :) You remember our call? We thought that ASP.NET Core adds mappings after the IServericeProvider was created. But that was wrong, because the ServiceProvider is <s>unmutable</s> immutable.

File: LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs (hunk 3 above)
Comment (JuergenGutsch) [reply LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs:15]: Nope, it wasn't my intent to have this argument to to enable unit testing. I tried to pass in the IHttpContextAccessor at initialization time. That was wrong, because the HttpContext is null. This is why I now create the instance of the HttpContextAccessor directly.
But it would make sense to add an internal constructor with the IHttpContextAccessor as argument, to enable unit testing here.
BTW: This is the ASP.NET Core HttpRequestLifecycle created by me, it is not the ASP.NET HttpRequestLifecycle created by you ;-)

File: LightCore/Registration/RegistrationSource/EnumerableRegistrationSource.cs (hunk 4 above)
Comment (JuergenGutsch) [reply LightCore/Registration/RegistrationSource/EnumerableRegistrationSource.cs:35]: Exactly :)

File: Samples/LightCore.Integration.AspNetCore.Sample/Startup.cs (hunk 5 above)
Comment (JuergenGutsch) [reply Samples/LightCore.Integration.AspNetCore.Sample/Startup.cs]: Tell MS not to add telemetry to ASP.NET Core by default ;-) Just kidding.
I'll check this. Maybe I will add this to the ASP.NET Core LightCore registration as a short term fix.
Just to get the preview 1 released...

File: Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md (hunk 1 above)
Comment (peterbucher) [reply Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md:4]: Yes i do remember. You mean "immutable"? OK.

File: LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs (hunk 3 above)
Comment (peterbucher) [reply LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs:15]: Ah okay. Yes you are right :-). The change makes sense now. And the propsal of adding a internal Constructor sounds good!

File: LightCore/Registration/RegistrationSource/EnumerableRegistrationSource.cs (hunk 4 above)
Comment (JuergenGutsch) [reply LightCore/Registration/RegistrationSource/EnumerableRegistrationSource.cs:35]: The actual fix is done in the ReceiveInstanceInLifecycle, where I just needed to rearrange the code a little bit, to get an empty list returned in case of missing registrations.

File: LightCore.Integration.AspNetCore/LightCoreExtension.cs (hunk 2 above)
Comment (JuergenGutsch) [reply LightCore.Integration.AspNetCore/LightCoreExtension.cs]: nothing. just code to remove. I added it to easily get some more information during debugging.
I'l remove it

**Comment by JuergenGutsch** [comment]
Hi @peterbucher 
many thanks for the review. 
I just anssered your questions in the comments.

Cheers

File: LightCore.Integration.AspNetCore/LightCoreExtension.cs (hunk 2 above)
Comment (peterbucher) [reply LightCore.Integration.AspNetCore/LightCoreExtension.cs]: OK!

Comment (JuergenGutsch) [reply LightCore.Integration.AspNetCore/LightCoreExtension.cs]: just code to remove

File: Samples/LightCore.Integration.AspNetCore.Sample/Startup.cs (hunk 5 above)
Comment (peterbucher) [reply Samples/LightCore.Integration.AspNetCore.Sample/Startup.cs]: Allright, but please hold it in mind for later fixing it properly.

File: LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs (hunk 3 above)
Comment (JuergenGutsch) [reply LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs:15]: 👍

File: Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md (hunk 1 above)
Comment (JuergenGutsch) [reply Samples/__LightCore.Integration.AspNetCore.Sample/Issues.md:4]: exactly :-)

**Review by peterbucher — APPROVED** [review]
Thanks @JuergenGutsch looks good now!

**Comment by peterbucher** [comment]
@JuergenGutsch Merged :-)

File: LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs (hunk 3 above)
Comment (JuergenGutsch) [reply LightCore.Integration.AspNetCore/Lifecycle/HttpRequestLifecycle.cs:15]: done

**Comment by JuergenGutsch** [comment]
Thanks @peterbucher :-)
